src/ripple/basics/impl/PerfLogImp.h

```diff
@@ -0,0 +1,250 @@
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_PERFLOGIMP_H
+#define RIPPLE_BASICS_PERFLOGIMP_H
+
+#include <ripple/basics/PerfLog.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/core/Stoppable.h>
+#include <ripple/app/main/Application.h>
+#include <ripple/protocol/JsonFields.h>
+#include <ripple/beast/core/SystemStats.h>
+#include <cstdint>
+#include <mutex>
+#include <memory>
+#include <unordered_map>
+#include <string>
+#include <utility>
+#include <tuple>
+#include <stdexcept>
+#include <vector>
+#include <fstream>
+#include <thread>
+#include <condition_variable>
+
+namespace ripple {
+namespace perf {
+
+/**
+ * Implementation class for PerfLog.
+ */
+class PerfLogImp
+    : public PerfLog, Stoppable
+{
+    /**
+     * Track performance counters and currently executing tasks.
+     */
+    class Counters
+    {
+    public:
+        /**
+         * RPC performance counters.
+         */
+        struct Rpc
+        {
+            std::uint64_t started {0};
+            std::uint64_t finished {0};
+            std::uint64_t errored {0};
+            std::uint64_t duration {0};
+            std::string label;
+            std::unique_ptr<Json::StaticString> labelSs;
+            mutable std::mutex mut;
+
+            Rpc() = default;
+
+            Rpc(Rpc const& orig)
+                : started (orig.started)
+                , finished (orig.finished)
+                , duration (orig.duration)
+                , label (orig.label)
+            {
+                if (orig.labelSs)
+                {
+                    labelSs = std::make_unique<Json::StaticString>(
+                        orig.labelSs->c_str());
+                }
+            }
+        };
+
+        /**
+         * Job Queue task performance counters.
+         */
+        struct Jq
+        {
+            std::uint64_t queued {0};
+            std::uint64_t started {0};
+            std::uint64_t finished {0};
+            std::uint64_t queuedDuration {0};
+            std::uint64_t runningDuration {0};
+            std::string label;
+            std::unique_ptr<Json::StaticString> labelSs;
```

Comment:
At first, I was going to say that this peculiar construct almost defeats the point of a `Json::StaticString` and would recommend using `char const* label = nullptr;` or, better still, just `Json::StaticString label`.

However, I think that this is broken and can result in dangling pointers. Remember: a `Json::StaticString` is a thin wrapper around a `char const*` and the thing it points to **MUST** remain valid as long as the `Json::StaticString` exists.

When we copy-construct a `Jq` which as a `labelSs` set, then we allocate memory for a new `Json::StaticString` and initialize its internal `char const*` to point to the same memory as the object we're going. And where does that memory come from? Ultimately, it points to the `label` in the parent object, which may have, long ago, gone out of scope.

There's simply no way to be sure that this can't happen without imposing onerous restrictions on this class _and_ then doing a careful review and lifetime analysis in every place where objects are stored and copied.

Are there any circumstances where `labelSs` points to a text that is different from what is in `label`? If not, my suggestion is to simply eliminate `labelSs` entirely and use `label` directly. If yes, then my suggestion is to add a second `std::string` and use it, as appropriate.

Reply:
@nbougalis Please see my fixes. I just constify'd those members and left a comment to never change them. Also fixed the copy constructors. The benefit of having the StaticString member there is so that it needn't be constructed every time the label is rendered.